core/threaded/scheduler_QS.c

```diff
@@ -0,0 +1,326 @@
+    // worker thread to become idle.
+    LF_PRINT_DEBUG(
+        "Scheduler: Worker %d is trying to acquire the scheduling "
+        "semaphore.",
+        worker_number);
+    lf_mutex_lock(&mutex);
+    if (++_lf_sched_instance->_lf_sched_number_of_idle_workers ==
+        _lf_sched_instance->_lf_sched_number_of_workers) {
+        // Last thread to go idle
+        LF_PRINT_DEBUG("Scheduler: Worker %d is the last idle thread.",
+                    worker_number);
+
+        // Nothing more happening at this tag.
+        LF_PRINT_DEBUG("Scheduler: Advancing tag.");
+        // This worker thread will take charge of advancing tag.
+        if (_lf_sched_advance_tag_locked()) {
+            LF_PRINT_DEBUG("Scheduler: Reached stop tag.");
+            _lf_sched_signal_stop();
+        }
+
+        // Reset all the PCs to 0.
+        for (int w = 0; w < _lf_sched_instance->_lf_sched_number_of_workers; w++) {
+            _lf_sched_instance->pc[w] = 0;
+        }
+        // Reset all the semaphores.
+        for (int i = 0; i < num_semaphores; i++) {
+            _lf_sched_instance->semaphores[i]->count = 0 ;
+        }
+
+        
+        _lf_sched_notify_workers();
+    } else {
+        // Not the last thread to become idle. Wait for work to be released.
+        // Wait for the last thread to signal the condition variable.
+        // lf_mutex_lock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        // lf_cond_wait(&_lf_sched_instance->_lf_sched_semaphore->cond, &_lf_sched_instance->_lf_sched_semaphore->mutex);
+        // lf_mutex_unlock(&_lf_sched_instance->_lf_sched_semaphore->mutex);
+        lf_cond_wait(&_lf_sched_instance->_lf_sched_array_of_conds[0], &mutex);
+        
+        LF_PRINT_DEBUG("Scheduler: Worker %d acquired the scheduling semaphore.",
+                    worker_number);
+    }
+    lf_mutex_unlock(&mutex);
+   // _lf_sched_notify_workers();
+}
+
+///////////////////// Scheduler Init and Destroy API /////////////////////////
+/**
+ * @brief Initialize the scheduler.
+ *
+ * This has to be called before other functions of the scheduler can be used.
+ * If the scheduler is already initialized, this will be a no-op.
+ *
+ * @param number_of_workers Indicate how many workers this scheduler will be
+ *  managing.
+ * @param option Pointer to a `sched_params_t` struct containing additional
+ *  scheduler parameters.
+ */
+void lf_sched_init(
+    size_t number_of_workers, 
+    sched_params_t* params
+) {
+    LF_PRINT_DEBUG("Scheduler: Initializing with %d workers", number_of_workers);
+
+    // This scheduler is unique in that it requires `num_reactions_per_level` to
+    // work correctly.
+    if (init_sched_instance(&_lf_sched_instance, number_of_workers, params)) {
+        // Scheduler has not been initialized before.
+        if (params == NULL) {
+            lf_print_error_and_exit(
+                "Scheduler: Internal error. The NP scheduler "
```

Comment:
Isn't this the QS scheduler, not the NP scheduler? Also, does params really need to be non-null? It is not used below.

Reply:
Yes, sorry about this. This is copied from the NP scheduler and should be removed.